File: routers/reportes.py
"""
SunQuote - Reportes API Router
Dashboard statistics and PDF data generation.
"""
from fastapi import APIRouter
from database import get_db

router = APIRouter()


@router.get("/dashboard/stats")
def dashboard_stats():
    """Get dashboard statistics."""
    conn = get_db()

    # Active quotes
    total_cotizaciones = conn.execute("SELECT COUNT(*) FROM cotizaciones").fetchone()[0]
    borradores = conn.execute("SELECT COUNT(*) FROM cotizaciones WHERE estado = 'borrador'").fetchone()[0]
    enviadas = conn.execute("SELECT COUNT(*) FROM cotizaciones WHERE estado = 'enviada'").fetchone()[0]
    firmadas = conn.execute("SELECT COUNT(*) FROM cotizaciones WHERE estado = 'firmada'").fetchone()[0]

    # Total revenue (from signed quotes)
    revenue_row = conn.execute("SELECT COALESCE(SUM(total_inversion), 0) FROM cotizaciones WHERE estado = 'firmada'").fetchone()
    total_revenue = revenue_row[0]

    # Total energy saved (sum of annual production from signed quotes)
    energy_row = conn.execute("SELECT COALESCE(SUM(produccion_mensual_kwh * 12), 0) FROM cotizaciones WHERE estado = 'firmada'").fetchone()
    total_energy_kwh = energy_row[0]

    # Conversion rate
    conversion = round((firmadas / total_cotizaciones * 100) if total_cotizaciones > 0 else 0, 1)

    # Total clients
    total_clientes = conn.execute("SELECT COUNT(*) FROM clientes").fetchone()[0]

    # Recent quotes
    recent = conn.execute("""
        SELECT c.id, c.codigo, c.estado, c.total_inversion, c.potencia_kwp, c.fecha,
               cl.nombre as cliente_nombre, cl.tipo_tarifa
        FROM cotizaciones c
        LEFT JOIN clientes cl ON c.cliente_id = cl.id
        ORDER BY c.fecha DESC LIMIT 5
    """).fetchall()

    recent_list = []
    for r in recent:
        recent_list.append({
            "id": r["id"],
            "codigo": r["codigo"],
            "estado": r["estado"],
            "total_inversion": r["total_inversion"],
            "potencia_kwp": r["potencia_kwp"],
            "fecha": r["fecha"],
            "cliente_nombre": r["cliente_nombre"],
            "tipo_tarifa": r["tipo_tarifa"]
        })

    conn.close()

    return {
        "total_cotizaciones": total_cotizaciones,
        "borradores": borradores,
        "enviadas": enviadas,
        "firmadas": firmadas,
        "total_revenue": total_revenue,
        "total_energy_kwh": total_energy_kwh,
        "conversion_rate": conversion,
        "total_clientes": total_clientes,
        "recent_cotizaciones": recent_list
    }

File: routes/reportes.js
const express = require('express');
const router = express.Router();
const pool = require('../db');

// ── Dashboard Stats ──
router.get('/dashboard/stats', async (req, res) => {
  try {
    const scope = '(empresa_id = ? OR ? = 1)';
    const scopeParams = [req.user.empresa_id, req.user.es_superadmin ? 1 : 0];
    const [[{total: totalCot}]] = await pool.execute(`SELECT COUNT(*) as total FROM cotizaciones WHERE ${scope}`, scopeParams);
    const [[{total: borradores}]] = await pool.execute(`SELECT COUNT(*) as total FROM cotizaciones WHERE estado = 'borrador' AND ${scope}`, scopeParams);
    const [[{total: enviadas}]] = await pool.execute(`SELECT COUNT(*) as total FROM cotizaciones WHERE estado = 'enviada' AND ${scope}`, scopeParams);
    const [[{total: firmadas}]] = await pool.execute(`SELECT COUNT(*) as total FROM cotizaciones WHERE estado = 'firmada' AND ${scope}`, scopeParams);
    const [[{total: revenue}]] = await pool.execute(`SELECT COALESCE(SUM(total_inversion), 0) as total FROM cotizaciones WHERE estado = 'firmada' AND ${scope}`, scopeParams);
    const [[{total: energy}]] = await pool.execute(`SELECT COALESCE(SUM(produccion_mensual_kwh * 12), 0) as total FROM cotizaciones WHERE estado = 'firmada' AND ${scope}`, scopeParams);
    const [[{total: totalClientes}]] = await pool.execute(`SELECT COUNT(*) as total FROM clientes WHERE ${scope}`, scopeParams);

    const conversion = totalCot > 0 ? Math.round(firmadas / totalCot * 1000) / 10 : 0;

    const [recent] = await pool.execute(`
      SELECT c.id, c.codigo, c.estado, c.total_inversion, c.potencia_kwp, c.fecha,
             cl.nombre as cliente_nombre, cl.tipo_tarifa
      FROM cotizaciones c
      LEFT JOIN clientes cl ON c.cliente_id = cl.id
      WHERE (c.empresa_id = ? OR ? = 1) ORDER BY c.fecha DESC LIMIT 5
    `, scopeParams);

    res.json({
      total_cotizaciones: totalCot,
      borradores,
      enviadas,
      firmadas,
      total_revenue: revenue,
      total_energy_kwh: energy,
      conversion_rate: conversion,
      total_clientes: totalClientes,
      recent_cotizaciones: recent
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ detail: 'Error del servidor' });
  }
});

module.exports = router;
